6-AdvancedScenarios/4-hybrid-SPA/ReadmeFiles/ReadmeAboutTheCode.md

```diff
@@ -0,0 +1,135 @@
+        .then((response) => {
+            res.json(response);
+        }).catch((error) => console.log(error))
+```
+
+Next, parse the authorization code, and invoke the acquireTokenByCode API on the ConfidentialClientApplication instance.
+
+When invoking this API, set enableSpaAuthorizationCode to true, which will enable MSAL to acquire a second authorization code to be redeemed by your single-page application.
+
+Your application should parse this second authorization code, as well as any account hints (e.g. sid, login_hint, preferred_username) and return them such that they can be rendered client-side:
+
+```javascript
+
+     const tokenRequest = {
+            code: req.body.code,
+            redirectUri: process.env.REDIRECT_URI,
+            enableSpaAuthorizationCode: true
+        };
+
+
+    
+    msalInstance.acquireTokenByCode(tokenRequest)
+        .then((response) => {
+
+            const { code } = response; //SPA authorization code
+            const {
+                sid, // Session ID claim, used for non-hybrid
+                login_hint: loginHint, // New login_hint claim (used instead of sid or email)
+                preferred_username: preferredUsername // Email
+            } = response.idTokenClaims;
+
+
+            req.session.code = code;
+            req.session.loginHint = loginHint;
+            req.session.sid = sid;
+            req.session.referredUsername = preferredUsername;
+            req.session.authenticated = true;
+
+            const urlFrom = (urlObject) => String(Object.assign(new URL("http://localhost:5000"), urlObject))
+            res.redirect(urlFrom({
+                 protocol: 'http',
+                 pathname: '/',
+                 search: 'getCode=true'
+            }))
+        }).catch((err) => {
+            console.log(err)
+        })
+```
+
+
+### Public client
+
+First, configure a new PublicClientApplication from MSAL.js in your single-page application:
+
+```javascript
+export const msalInstance = new PublicClientApplication(msalConfig);
+
+ReactDOM.render(
+  <BrowserRouter>
+    <App instance={msalInstance}/>
+  </BrowserRouter>,
+  document.getElementById('root')
+);
+```
+
+Next, render the code that was acquired server-side, and provide it to the acquireTokenByCode API on the MSAL.js PublicClientApplication instance.
+
+The application should also render any account hints, as they will be needed for any interactive requests to ensure the same user is used for both requests.
+
+```javascript
+if(getCode && !data){
+      callApiToGetSpaCode()
+        .then((response) => {
+        if(inProgress === "none"){
+          const { code, loginHint, sid, referredUsername } = response;
+          instance.acquireTokenByCode({
+            code
+          }).then((res) => {
+              setdata(res)
+            }).catch((error ) => {
+              if(error instanceof InteractionRequiredAuthError){
+                 if (inProgress === "none") {
```

Comment:
You should probably not make the call to `acquireTokenByCode` at all if `inProgress` is something other than `none`

Reply:
This format is for access token v1, and it's not applicable for access token v2. We instruct our users to update the access token to v2 in the sample doc.

Reply:
I am only calling `acquireTokenByCode` if the `inProgress ` is equal `none`. I believe that should fulfill the requirement; let me know if I missed something. 
```javascript
      let apiData;
      let token;

      if (getCode && !data) {
        apiData = await callApiToGetSpaCode();
        const { code, loginHint, sid, referredUsername } = apiData;

        if (inProgress === "none") {
          try {
            token = await instance.acquireTokenByCode({
              code, //Spa Auth code
            });

            setdata(token);
          } catch (error) {
            if (error instanceof InteractionRequiredAuthError) {
              if (inProgress === "none") {
                //If loginHint claim is provided, dont use sid
                try {
                  token = await instance.loginPopup({
                    loginHint, //Prefer loginHint claim over referredUsername (email)
                  });

                  setdata(token);
                } catch (error) {
                  console.log(error);
                }
              }
            }
          }
        }
      }
```

Reply:
The point is that I don't think you want/need the second `if (inProgress === "none") {` check inside the catch. @tnorling thoughts?

Reply:
In theory `inProgress` could change between when `acquireTokenByCode` is called and when it throws. It's usually a good idea to check it right before calling an interactive API, however, since `inProgress` is a state value the value won't be different between these two evaluations, if inProgress changes it will trigger a rerender and the whole code block will re-run with the new value on the new render. It may be better to use a ref here so that this doesn't get re-run on each change of `inProgress` and you can access the current value of `inProgress` when you need it.